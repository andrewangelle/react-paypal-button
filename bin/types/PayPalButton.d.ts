/// <reference types="react" />
import '@babel/polyfill';
import { PayPalButtonProps } from './types';
declare function PayPalButton(props: PayPalButtonProps): JSX.Element;
export default PayPalButton;
