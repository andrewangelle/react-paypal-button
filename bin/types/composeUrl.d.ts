declare function composeUrl(apiKey: string): string;
export default composeUrl;
