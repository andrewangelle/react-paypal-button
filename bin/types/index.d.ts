export { default as PayPalButton } from './PayPalButton';
export { PayPalButtonProps, PayPalPaymentData, OnShippingChangeData, OnCancelData, OnCaptureData, OnApproveData } from './types';
export { default as useExpressCheckout } from './useExpressCheckout';
