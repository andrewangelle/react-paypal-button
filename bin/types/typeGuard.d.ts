import { OnApproveData, OnCaptureData } from 'types';
export declare function typeGuard(str: string | undefined): string;
export declare function typeGuardOnApprove(data: OnApproveData | OnCaptureData): OnApproveData | OnCaptureData;
