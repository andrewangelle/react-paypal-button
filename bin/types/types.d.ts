declare global {
    interface Window {
        paypal: any;
    }
}
export declare type OnCancelData = {
    billingID: string;
    cancelUrl: string;
    intent: string;
    paymentID: string;
    paymentToken: string;
};
export declare type OnShippingChangeData = {
    amount: {
        value: string;
        currency_code: string;
        breakdown: {};
    };
    orderID: string;
    paymentID: string;
    paymentToken: string;
    shipping_address: {
        city: string;
        country_code: string;
        postal_code: string;
        state: string;
    };
};
export declare type OnApproveData = {
    orderID: string;
    payerID: string;
};
export declare type OnCaptureData = {
    create_time: string;
    id: string;
    intent: String;
    links: Array<{
        href: string;
        method: string;
        rel: string;
        title: string;
    }>;
    payer: {
        address: {
            country_code: string;
        };
        email_address: string;
        name: {
            given_name: string;
            surname: string;
        };
        payer_id: string;
    };
    purchase_units: Array<object>;
    status: string;
    update_time: string;
};
export declare type PayPalPaymentData = {
    cart: string;
    create_time: string;
    id: string;
    intent: 'sale' | 'purchase';
    payer: {
        payer_info: {
            country_code: string;
            email: string;
            first_name: string;
            last_name: string;
            middle_name: string;
            payer_id: string;
            shipping_address: {
                city: string;
                country_code: string;
                line1: string;
                postal_code: string;
                recipient_name: string;
                state: string;
            };
        };
        payment_method: string;
        status: 'UNVERIFIED' | 'VERIFIED';
    };
    state: string;
    transaction: any[];
};
declare type OnShippingChangeReturnType = Promise<number | void> | number | void;
export interface PayPalButtonProps {
    env: 'sandbox' | 'production';
    sandboxID?: string;
    productionID?: string;
    amount: number;
    currency: string;
    intent?: 'capture' | 'authorize';
    onApprove?: (data: {
        orderID: string;
        payerID: string;
    }, authId: string) => void;
    onPaymentStart?: () => void;
    onPaymentSuccess?: (response: PayPalPaymentData | OnCaptureData) => void;
    onPaymentError?: (msg: string) => void;
    onPaymentCancel?: (data: OnCancelData) => void;
    onShippingChange?: (data: OnShippingChangeData) => OnShippingChangeReturnType;
}
export interface PaypalOptions {
    clientId?: string;
    merchantId?: string;
    currency?: number | string;
    intent?: string;
    commit?: boolean | string;
    vault?: boolean | string;
    component?: string;
    disableFunding?: string;
    disableCard?: string;
    integrationDate?: string;
    locale?: string;
    buyerCountry?: string;
    debug?: boolean | string;
}
export {};
