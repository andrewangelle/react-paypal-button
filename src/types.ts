
declare global {
  interface Window {
    paypal: any
  }
}

export type OnCancelData = {
  billingID: string;
  cancelUrl: string;
  intent: string;
  paymentID: string;
  paymentToken: string;
}

export type OnShippingChangeData = {
  amount: {
    value: string,
    currency_code: string,
    breakdown: {}
  },
  orderID: string;
  paymentID: string;
  paymentToken: string;
  shipping_address: {
    city: string;
    country_code: string;
    postal_code: string;
    state: string;
  }
}

export type OnApproveData = {
  orderID: string,
  payerID: string
};

export type OnCaptureData = {
  create_time: string;
  id: string;
  intent: String;
  links: Array<{href: string; method: string; rel: string; title: string;}>
  payer: {
    address: {country_code: string}
    email_address: string;
    name: {
      given_name: string; // first name
      surname: string; // last name
    },
    payer_id: string;
  };
  purchase_units: Array<object>;
  status: string;
  update_time: string;
}


type OnShippingChangeReturnType =
  Promise<number | void> |
  number |
  void;

export type PaypalOptions = {
  clientId: string,
  merchantId?: string,
  currency?: number|string,
  intent: 'capture' | 'authorize',
  commit?: boolean|string,
  vault?: boolean|string,
  components?: string,
  disableFunding?: string,
  disableCard?: string, // card, credit, bancontact
  integrationDate?: string,
  locale?: string,
  buyerCountry?: string,
  debug?: boolean|string
}

export type ButtonStylingOptions = {
  layout?:  'vertical' | 'horizontal',
  color?:   'blue' | 'gold' | 'silver' | 'white' | 'black';
  shape?:   'rect' | 'pill';
  label?:   'paypal' | 'checkout' | 'buynow' | 'pay' | 'installment';
  tagline?: boolean;
}

export interface PayPalButtonProps {
  paypalOptions: PaypalOptions;
  buttonStyles?: ButtonStylingOptions,
  amount: number | string;
  onApprove?: (data: OnApproveData, authId: string) => void;
  onPaymentStart?: () => void;
  onPaymentSuccess?: (response: OnCaptureData) => void;
  onPaymentError?: (msg: string) => void;
  onPaymentCancel?: (data: OnCancelData) => void;
  onShippingChange?: (data: OnShippingChangeData) => OnShippingChangeReturnType;
}

