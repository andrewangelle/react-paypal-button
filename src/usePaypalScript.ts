import { useState, useEffect } from 'react';
import { PayPalButtonProps } from './types';
import composeUrl from './composeUrl';

type Props = {
  loading: boolean;
  done: boolean;
  error:boolean;
}

function usePaypalScript(props: PayPalButtonProps): Props {
  let scriptCache: string[] = [];

  const url = composeUrl({
    clientId: props.sandboxID || props.productionID,
    intent: props.intent
  });

  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [done, setDone] = useState(false);


  useEffect(() => {
    if(scriptCache.includes(url)){
      setLoading(false)
      setDone(true),
      setError(false)
    } else {
      scriptCache.push(url)
    }

    let script = document.createElement('script')
    script.src = url;
    script.async = true;

    const onScriptLoad = () => {
      setLoading(false)
      setDone(true),
      setError(false)
    }

    const onScriptError = () => {
      // if we error out, retry by removing the url from the cache
      const urlIndex = scriptCache.indexOf(url);
      if(urlIndex !== -1){
        scriptCache.splice(urlIndex, 1);
        script.remove()
      }
      setLoading(false)
      setDone(true),
      setError(true)
    }

    script.addEventListener('load', onScriptLoad);
    script.addEventListener('error', onScriptError);

    document.body.appendChild(script)

    return () => {
      script.removeEventListener('load', onScriptLoad);
      script.removeEventListener('error', onScriptError);
    }
  }, [url]);

  return { loading, error, done }
}

export default usePaypalScript