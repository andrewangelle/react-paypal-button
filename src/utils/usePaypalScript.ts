import { useState, useEffect } from 'react';

import { PaypalOptions } from '../types';
import { scriptLoadError, composeUrl } from '.';

type Props = {
  loading: boolean;
  done: boolean;
  error:boolean;
}

export function usePaypalScript(options: PaypalOptions): Props {
  let scriptCache: string[] = [];

  const url = composeUrl(options);

  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [done, setDone] = useState(false);


  useEffect(() => {
    // early exit if cache is in script
    if(scriptCache.includes(url)){
      setLoading(false)
      setDone(true),
      setError(false)
    } else {
      scriptCache.push(url)
    }

    let script = document.createElement('script')
    script.src = url;
    script.async = true;

    const onScriptLoad = () => {
      setLoading(false)
      setDone(true),
      setError(false)
    }

    const onScriptError = () => {
      // if we error out, retry by removing the url from the cache
      const urlIndex = scriptCache.indexOf(url);

      if(urlIndex !== -1){
        scriptCache.splice(urlIndex, 1);
        script.remove()
      }

      console.error(scriptLoadError)

      setLoading(false)
      setDone(true),
      setError(true)
    }

    script.addEventListener('load', onScriptLoad);
    script.addEventListener('error', onScriptError);

    document.body.appendChild(script)

    return () => {
      script.removeEventListener('load', onScriptLoad);
      script.removeEventListener('error', onScriptError);
    }

    // rerun if url changes
  }, [url]);

  return { loading, error, done }
}
