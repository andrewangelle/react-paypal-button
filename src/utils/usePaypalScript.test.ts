import { renderHook } from '@testing-library/react-hooks';
import { usePaypalScript } from './usePaypalScript';
import { PaypalOptions } from '../types';

const props: PaypalOptions = {
  clientId: '12345',
  intent: 'capture'
}

describe('usePaypalMethods', () => {
  it('returns loading state', () => {
    const { result }= renderHook(() => usePaypalScript(props))
    expect(result.current).toMatchSnapshot()
  })
})