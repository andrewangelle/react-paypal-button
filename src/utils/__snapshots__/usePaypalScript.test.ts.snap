// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePaypalMethods returns loading state 1`] = `
Object {
  "done": false,
  "error": false,
  "loading": true,
}
`;
